app/routes/devices.$id.hardware.$hardware_name.calibrate.tsx

```diff
@@ -0,0 +1,252 @@
+        },
+        client: evolverClient,
+      },
+    );
+
+  return json({ actions: data?.actions });
+}
+
+const CalibrationProcedure = ({ actions }) => {
+  const submit = useSubmit();
+  const { id, hardware_name } = useParams();
+
+  return actions.map((action, ix) => {
+    const dispatchAction = (actionFormData: object) => {
+      const formData = new FormData();
+      formData.append("id", id ?? "");
+      formData.append("intent", Intent.Enum.dispatch_action);
+      formData.append("hardware_name", hardware_name ?? "");
+      formData.append("action_name", action.name);
+      formData.append("payload", JSON.stringify(actionFormData));
+      submit(formData, {
+        method: "POST",
+      });
+    };
+    return (
+      <ClientOnly
+        key={action.description}
+        fallback={<span className="skeleton h-32"></span>}
+      >
+        {() => (
+          <CalibratorActionForm
+            action={action}
+            index={ix}
+            dispatchAction={dispatchAction}
+          />
+        )}
+      </ClientOnly>
+    );
+  });
+};
+
+export default function CalibrateHardware() {
+  const { actions } = useLoaderData<typeof loader>();
+  const { id, hardware_name } = useParams();
+  const hasActions = actions && actions.length > 0;
+  const calibrationButtonCopy = hasActions ? "restart" : "start";
+
+  const actionData = useActionData<typeof action>();
+  const submit = useSubmit();
+
+  useEffect(() => {
+    if (actionData?.error) {
+      if (typeof actionData.error === "string") {
+        notify.error(actionData.error);
+      }
+      if (typeof actionData.error === "object") {
+        const errorMessages: string[] = [];
+        Object.entries(actionData.error).forEach(([key, value]) => {
+          errorMessages.push(`${key}: ${value}`);
+        });
+        errorMessages.forEach((message) => {
+          notify.error(message);
+        });
+      }
+    }
+  }, [actionData]);
+
+  return (
+    <div>
+      <div className="flex flex-col gap-4">
+        <div className="flex justify-between items-center gap-4">
+          <div className="flex items-center">
+            <div>
+              <div className="flex w-full font-mono">calibration procedure</div>
+            </div>
+          </div>
+          <div>
+            <button
+              className={clsx(
+                "btn",
+                hasActions && "btn-error",
+                !hasActions && "btn-accent",
+              )}
+              onClick={() => {
+                const formData = new FormData();
+                formData.append("id", id ?? "");
+                formData.append(
+                  "intent",
+                  Intent.Enum.start_calibration_procedure,
+                );
+                formData.append("hardware_name", hardware_name ?? "");
+                submit(formData, {
+                  method: "POST",
+                });
+              }}
+            >
+              {calibrationButtonCopy}
```

Comment:
I don't see any particular feedback when clicking restart, form doesn't clear. It might be that when the action buttons toggle it will be visible, but also thinking that if "restart" is destructive, we might want some kind of confirmation and info to the user what would happen.

Reply:
Yep this also depends on the backend change mentioned above. Adding a confirmation dialog is a good idea as it is destructive.